cmd/kubeadm/app/phases/upgrade/prepull.go

```diff
@@ -0,0 +1,177 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrade
+
+import (
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+)
+
+const (
+	prepullPrefix = "upgrade-prepull-"
+)
+
+type prepuller interface {
+	CreateFunc(string)
+	WaitFunc(string)
+	DeleteFunc(string)
+}
+
+type DaemonSetPrepuller struct {
+	client clientset.Interface
+	cfg    *kubeadmapi.MasterConfiguration
+}
+
+func NewDaemonSetPrepuller(client clientset.Interface, cfg *kubeadmapi.MasterConfiguration) *DaemonSetPrepuller {
+	return &DaemonSetPrepuller{
+		client: client,
+		cfg:    cfg,
+	}
+}
+
+// CreateFunc creates a DaemonSet for making the image available on every relevant node
+func (d *DaemonSetPrepuller) CreateFunc(component string) {
+	ds := buildPrePullDaemonSet(d.cfg, component)
+
+	// Create the DaemonSet in the API Server
+	if _, err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Create(ds); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			fmt.Printf("[upgrade/prepull] WARNING: Failed to create daemonset for prepulling component %q: %v", component, err)
+			return
+		}
+
+		if _, err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Update(ds); err != nil {
+			// TODO: We should retry on 409 responses
+			fmt.Printf("[upgrade/prepull] WARNING: Failed to update daemonset for prepulling component %q: %v", component, err)
+		}
+	}
+}
+
+// WaitFunc waits for all Pods in the specified DaemonSet to be in the Running state
+func (d *DaemonSetPrepuller) WaitFunc(component string) {
+	fmt.Printf("[upgrade/prepull] Prepulling image for component %s.\n", component)
+	kubeadmutil.WaitForPodsWithLabel(d.client, 0, ioutil.Discard, "k8s-app=upgrade-prepull-"+component)
+}
+
+// DeleteFunc deletes the DaemonSet used for making the image available on every relevant node
+func (d *DaemonSetPrepuller) DeleteFunc(component string) {
+	dsName := addPrepullPrefix(component)
+	foregroundDelete := metav1.DeletePropagationForeground
+	deleteOptions := &metav1.DeleteOptions{
+		PropagationPolicy: &foregroundDelete,
+	}
+	if err := d.client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Delete(dsName, deleteOptions); err != nil {
```

Comment:
should we return an error here? In other words, should a failed cleanup halt execution?